compiler/src/iree/compiler/Preprocessing/Common/PadToIntrinsics.cpp

```diff
@@ -30,26 +31,116 @@ namespace mlir::iree_compiler::Preprocessing {
-  auto high = llvm::map_to_vector(padding, [&](int64_t v) -> OpFoldResult {
-    return rewriter.getIndexAttr(v);
-  });
   Value paddedResult = rewriter.create<tensor::PadOp>(
-      loc, paddedResultType, padSource, low, high, paddingValue);
+      loc, paddedResultType, padSource, low, padding, paddingValue);
   return paddedResult;
 }
 
+static Value
+getExpandedValue(RewriterBase &rewriter, Location loc, Value expandSource,
+                 AffineMap &operandMap,
+                 SmallVector<std::pair<int64_t, int64_t>> &dimsToExpand) {
+  auto srcType = expandSource.getType().cast<RankedTensorType>();
+  auto srcShape = srcType.getShape();
+  SetVector<int64_t> operandDimsToExpand;
+  SmallVector<std::optional<int64_t>> operandDimToExpandSize(srcType.getRank(),
+                                                             std::nullopt);
+  // Get dims to expand from operand's POV.
+  for (auto [dimToExpand, sizeToExpand] : dimsToExpand) {
+    std::optional<int64_t> maybeDim = operandMap.getResultPosition(
+        getAffineDimExpr(dimToExpand, operandMap.getContext()));
+    if (maybeDim) {
+      operandDimsToExpand.insert(maybeDim.value());
+      operandDimToExpandSize[maybeDim.value()] = sizeToExpand;
+    }
+  }
+  if (operandDimsToExpand.empty()) {
+    return expandSource;
+  }
+
+  // Form reassocation indices and expanded shape.
+  SmallVector<ReassociationIndices> reassoc;
+  SmallVector<int64_t> expandedShape;
+  int64_t reassocOffset = 0;
+  for (int i = 0; i < srcType.getRank(); i++) {
+    if (operandDimToExpandSize[i]) {
+      expandedShape.append({srcShape[i], operandDimToExpandSize[i].value()});
+      reassoc.push_back(
+          ReassociationIndices{reassocOffset + i, reassocOffset + i + 1});
+      ++reassocOffset;
+    } else {
+      expandedShape.push_back(srcShape[i]);
+      reassoc.push_back(ReassociationIndices{reassocOffset + i});
+    }
+  }
+
+  return rewriter.create<tensor::ExpandShapeOp>(
+      loc, RankedTensorType::Builder(srcType).setShape(expandedShape),
+      expandSource, reassoc);
+}
+
+// expandMapsAndIterators expands the maps and iterators of an linalgOp at
+// dimensions specified by `dimsToExpand`. The general idea is we want to view
+// the dimensions specified in `dimsToExpand` as being splitted-up to two
+// dimensions. Which means the src dim and the new child/sub dim will be
+// contiguous.
+
+// For example if we have:
+// affine_map<(d0, d1, d2) -> (d0, d2, d1)>
+// iterator_types = ["parallel, "parallel", "reduce"]
+// dimsToExpand = [1, 2]
+
+// We will turn this into:
+// affine_map<(d0, d1_A, d1_B, d2_A, d2_B) -> (d0, d2_A, d2_B, d1_A, d1_B)>
+// iterator_types = ["parallel, "parallel", "parallel", "reduce", "reduce"]
+static void
+expandMapsAndIterators(SmallVector<AffineMap> &expandedMaps,
+                       SmallVector<utils::IteratorType> &expandedIterators,
+                       SmallVector<std::pair<int64_t, int64_t>> &dimsToExpand) {
+  int expandOffset = 0;
```

Comment:
Could this logic be made simpler with https://github.com/shark-infra/llvm-project/blob/c9866736ccbd55026af03c8e3ceee41a6186d63c/mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp#L32

Reply:
Hey Mahesh, I think with we can use `getReassociationIndicesForReshape` to get some metadata, and would be able to query the expanded dims from reassociation indices instead of using:
```c
      int64_t expandSrcDim = expandDim + expandOffset;
      int64_t expandDstDim = expandSrcDim + 1;
```

This would simplify the dim tracking and offset-dimension a bit. But this may add extra complexity in logic since reassociation indices is dimension from operand POV, while we need to also handle iteratorTypes from the linalgOp's POV and shuttling of srcShape and targetShape of the operands. Any thoughts?

Reply:
Ok, I am fine with leaving this as is.